### DISCUSSION TITLE
Does ONNXRuntime support YOLOv11 .onnx files?

### DISCUSSION THREAD

**Discussion by vadimax** [microsoft/onnxruntime#22298] [General] [answered]
Hello everyone,

I try to switch from YOLOv10 to YOLOv11 and fail. The reason is unknown (read weird) information being returned by session.Run() method (C++). When I retrieve output tensor shape information:

Ort::TensorTypeAndShapeInfo tensor_info=out_tensors.front().GetTensorTypeAndShapeInfo();
if((out_number=tensor_info.GetDimensionsCount())!=3)
	throw Status("Output tensor number of shape dimensions is %llu instead of 3",out_number);
intermediate=tensor_info.GetShape();
out_shape.batch=intermediate[0];
out_shape.detections=intermediate[1];
out_shape.attributes=intermediate[2];

I get weird results:

out_shape.batch gets 1 -- (ok);
out_shape.detections gets 84 -- (well, perhaps, ok);
out_shape.attributes gets 8400 -- (what the hell is THAT?).

When I was using YOLOv10?.onnx the third parameter was output tensor data size (6: left, top, right, bottom, class_id, confidence). With YOLOv11?.onnx it is 8400. I guess there is a compatibility issue or something like that.

Kind regards, Vadim.

**Comment by Neakita** [comment]
There is no any compatibility issue. I am using Yolo V8 right now and it is also have 8400 shape at 640x640 image size.

https://readmedium.com/how-to-get-bounding-boxes-and-class-probabilities-from-yolov5-and-yolov8-raw-output-4c4d3dd93fae

> To be precise, the three detection heads each progressively use a stride (the scaling factore at which each head sees the input image) of 8, 16, 32.
> 
> In other words,
> 
> First head uses a stride of 8 and will produce an output map of 640/8 = 80 (80x80 pixels) = 6400 pixels
> 
> Second head uses a stride of 16 and will produce an output map of 640/16 = 40 pixels (40x40) = 1600 pixels
> 
> Third head uses a stride of 32 and will similarly produce an output map of 640/ 32 = 20 pixels (20x20) = 400 pixels
> 
> Thus, we have 6400 + 1600 + 400 = 8400 pixels in the output feature map.
> 
> This accounts for the 8400 in the (1, 84, 8400) of the output shape.

**Reply by vadimax** [reply]
This is still weird. If output data block looks like

```
struct DATABLOCK
          {
          float x,y,w,h,confidence[80];
          };
```

then confidence values should stay in [0.0...1.0] range, but they go beyond 600.

![Weird confidence values](https://github.com/user-attachments/assets/1a2eb5f3-f1be-47cb-9732-17fd8cca1d67)

**Reply by Neakita** [reply]
i just exported v11 to onnx and used parser from v8, detection model output parsed just fine, so it's probably same output shape as v8.
edit: my confidence values in normalized range, just as it should be.

**Comment by vadimax** [comment]
May anyone describe the output data format? Because sample code variants are:

1. {x,y,w,h,confidence[80]} — does not work;
2. anything else?

YOLOv10?.onnx I used before produced {left,top,right,bottom,class_id,confidence}.

**Reply by Neakita** [reply]
See [this](https://github.com/dme-compunet/YoloV8/blob/87c12f0795a3c359118a3f9efb744c8bde0b4688/Source/YoloV8/Services/Parsers/RawBoundingBoxParser.cs#L8) or [this](https://github.com/Neakita/Yolo/blob/4f95871cfba2ea01ba3197e185b6d71ecc164080/Yolo/OutputProcessing/V8DetectionProcessor.cs#L30)

**Comment by vadimax** [comment, marked as answer]
Holy crap! Now I understand what is going on. You need to transpose the output matrix to get human readable format:

{ x, y, w, h, confidence[80] }

YOLO v5, v7, v10 use human readable output format;
YOLO v8, v11 -- its transposition.

WHY?!
